lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,305 @@
+      @has_entered = true
+      raise RuntimeError, "TruffleRuby does not support call after stop" if @stopped
+      begin
+        translate do
+          function = eval_in_context(function_name)
+          function.call(*convert_ruby_to_js(arguments))
+        end
+      rescue RuntimeError => e
+        raise e unless e.message == "Polyglot::InnerContext was terminated forcefully"
+      rescue SystemStackError => e
+        raise RuntimeError, e.message
+      end
+    ensure
+      @entered = false
+    end
+
+    def create_isolate_value
+      # Returning a dummy object since TruffleRuby does not have a 1-1 concept with isolate.
+      # However, code and ASTs are shared between contexts.
+      Isolate.new
+    end
+
+    def isolate_mutex
+      @isolate_mutex
+    end
+
+    class ExternalFunction
+      private
+
+      def notify_v8
+        name = @name.encode(::Encoding::UTF_8)
+        wrapped = lambda do |*args|
+          converted = @parent.send(:convert_js_to_ruby, args)
+          @parent.send(:convert_ruby_to_js, @callback.call(*converted))
+        end
+
+        if @parent_object.nil?
+          # set global name to proc
+          result = @parent.eval_in_context('this')
+          result[name] = wrapped
+        else
+          parent_object_eval = @parent_object_eval.encode(::Encoding::UTF_8)
+          result = @parent.eval_in_context(parent_object_eval)
+          result[name] = wrapped
+          # set evaluated object results name to proc
+        end
+      end
+    end
+
+    def translate
+      convert_js_to_ruby yield
+    rescue ::RuntimeError => e
+      if e.message.start_with?('SyntaxError:')
+        error_class = MiniRacer::ParseError
+      else
+        error_class = MiniRacer::RuntimeError
+      end
+
+      backtrace = e.backtrace.map { |line| line.sub('(eval)', '(mini_racer)') }
+      raise error_class, e.message, backtrace
+    end
+
+    def convert_js_to_ruby(value)
+      case value
+      when true, false, Integer, Float
+        value
+      else
+        if value.nil?
+          nil
+        elsif value.respond_to?(:call)
+          MiniRacer::JavaScriptFunction.new
+        elsif value.respond_to?(:to_str)
+          value.to_str
+        elsif value.respond_to?(:to_ary)
+          value.to_ary.map do |e|
+            if e.respond_to?(:call)
+              nil
+            else
+              convert_js_to_ruby(e)
+            end
+          end
+        elsif is_time(value)
+          js_date_to_time(value)
+        elsif is_symbol(value)
+          js_symbol_to_symbol(value)
+        else
+          object = value
+          h = {}
+          object.instance_variables.each do |member|
+            v = object[member]
+            unless v.respond_to?(:call)
+              h[member.to_s] = convert_js_to_ruby(v)
+            end
+          end
+          h
+        end
+      end
+    end
+
+    def is_time(value)
+      f = eval_in_context("(x) => { return x instanceof Date };")
```

Comment:
Such functions should be stored as `@ivars` in the constructor so we don't repeatedly parse them for a given context

Reply:
I used a pattern like this in the helper methods to avoid eval'ing these eagerly in the constructor:
```
@is_time_func ||= eval_in_context("(x) => { return x instanceof Date };")
```